llvm/lib/Support/Windows/Path.inc

```diff
@@ -1373,29 +1374,10 @@ std::error_code closeFile(file_t &F) {
 }
 
 std::error_code remove_directories(const Twine &path, bool IgnoreErrors) {
-  // Convert to utf-16.
-  SmallVector<wchar_t, 128> Path16;
-  std::error_code EC = widenPath(path, Path16);
-  if (EC && !IgnoreErrors)
-    return EC;
-
-  // SHFileOperation() accepts a list of paths, and so must be double null-
-  // terminated to indicate the end of the list.  The buffer is already null
-  // terminated, but since that null character is not considered part of the
-  // vector's size, pushing another one will just consume that byte.  So we
-  // need to push 2 null terminators.
-  Path16.push_back(0);
-  Path16.push_back(0);
-
-  SHFILEOPSTRUCTW shfos = {};
-  shfos.wFunc = FO_DELETE;
-  shfos.pFrom = Path16.data();
-  shfos.fFlags = FOF_NO_UI;
-
-  int result = ::SHFileOperationW(&shfos);
-  if (result != 0 && !IgnoreErrors)
-    return mapWindowsError(result);
-  return std::error_code();
+  const std::filesystem::path Path(path.str());
```

Comment:
Unfortunately, passing paths like this won't do the right thing.

Within LLVM, paths in the usual form with narrow chars is assumed to be UTF-8; for the existing Windows implementations here, it is converted to UTF-16 into `wchar_t` form with `widenPath` above.

However, in `std::filesystem`, a narrow char path is assumed to be in the system active code page (ACP). Unless the ACP is set to UTF-8, only plain ASCII path names would work correctly.

There is a `std::filesystem::u8path()` function which takes `char` input and interprets it as UTF-8, which we could use instead. That one is deprecated since C++20 though (because in C++20, there's a `char8_t` data type to natively signal UTF8 strings).

I guess it's simplest to just keep the existing `widenPath` call first and then pass the path from that to `std::filesystem` though, but I don't have a very strong opinion on it - `u8path()` might be ok.


(At least this is how MS STL and libc++ implements `std::filesystem`. I haven't observed libstdc++'s implementation much at all; I'm not sure how mature it is on Windows.)

Reply:
MS STL calls
https://github.com/microsoft/STL/blob/89ca073a866d24db42b3f4c5e406be4b28a148f4/stl/inc/filesystem#L655
```
template <class _Src, enable_if_t<_Is_Source<_Src>, int> = 0>
path(const _Src& _Source, format = auto_format) : _Text(_Convert_Source_to_wide(_Source)) {
    // format has no meaning for this implementation, as the generic grammar is acceptable as a native path
}
```
https://github.com/microsoft/STL/blob/89ca073a866d24db42b3f4c5e406be4b28a148f4/stl/inc/filesystem#L280
```
template <class _Src, class _Conversion = _Normal_conversion>
_NODISCARD wstring _Convert_Source_to_wide(const _Src& _Source, _Conversion _Tag = {}) {
    return _Convert_stringoid_to_wide(_Stringoid_from_Source(_Source), _Tag);
}
```
https://github.com/microsoft/STL/blob/89ca073a866d24db42b3f4c5e406be4b28a148f4/stl/inc/filesystem#L193-L201
```
template <class _Conversion>
_NODISCARD wstring _Convert_stringoid_to_wide(const string_view _Input, _Conversion) {
    _STL_INTERNAL_STATIC_ASSERT(_Is_any_of_v<_Conversion, _Normal_conversion, _Utf8_conversion>);
     if constexpr (is_same_v<_Conversion, _Normal_conversion>) {
        return _Convert_narrow_to_wide(__std_fs_code_page(), _Input);
    } else {
        return _Convert_narrow_to_wide(__std_code_page::_Utf8, _Input);
    }
}
```
and `__std_fs_code_page()` returns CP_ACP if system's locale is English.
So, yes, it does not expect any non-ASCII symbols in the path.
I have updated the patch to use widenPath() and tested it with non-ASCII symbols locally.